Draft EMTFShower producer

dildick opened the pull request:
#### PR description:

Draft of an EMTFShower producer. It is not meant to be merged, but serve as a starting point for a discussion on the architecture.

@rekovic I added a few bits and pieces for

1) CSC local trigger (hit counting + bit assignment in the trigger data format)
2) ETMF: new object EMTFShower and new class ShowerRecognition in the EMTF sector processor. EMTFShowers would be sent to uGMT
3) uGMT: receives EMTFShowers from 12 sectors. Does simple operation (likely just an OR). Send a single bit or a few bits to uGT to indicate presence of a showers. uGMT produces l1t::Shower which contains the relevant bits.

I kept the idea rather general, assuming that in Run-4 the BMTF and OMTF would also produce showers. 

#### PR validation:

Code compiles. No validation done yet.

#### if this PR is a backport please specify the original PR:

N/A

dildick commented:
FYI @dinyar

dinyar commented on DataFormats/L1TMuon/interface/EMTFShower.h:
```diff
@@ -0,0 +1,32 @@
+    int sector() const { return sector_; }
+
+  private:
+    unsigned int bits_;
```
I'm not a big friend of storing unstructured information in data formats. I suspect you want to do a quick OR of the return values of `bits()` in the uGMT emulator, but I think you can achieve this by storing and exposing the fields implicitly containted in `bits_` and then having a utility function that computes whatever `bits()` returned. For the Muon data formats we have some similar utility function in e.g. the MuonRawToDigiTranslator.

dildick replied:
okay, I think bits_ was template until we came up with a first draft of the EMTF data format. We can expose this to make it clear to the user.

dinyar commented on DataFormats/L1TMuon/src/classes_def.xml:
```diff
@@ -38,6 +38,9 @@
   <class name="l1t::EMTFTrackCollection"/>
   <class name="edm::Wrapper<l1t::EMTFTrackCollection>"/>
   <class name="l1t::EMTFPtLUT"/>
+  <class name="l1t::EMTFShower"/>
```
Should we aim to version `EMTFShower` from the beginning?

dildick replied:
Not sure. Maybe l1t::Shower is better?

dinyar commented on L1Trigger/L1TMuonEndCap/plugins/L1TMuonEndCapTrackProducer.cc:
```diff
@@ -8,6 +8,7 @@ L1TMuonEndCapTrackProducer::L1TMuonEndCapTrackProducer(const edm::ParameterSet&
   produces<EMTFHitCollection>("");                      // All CSC LCTs and RPC clusters received by EMTF
   produces<EMTFTrackCollection>("");                    // All output EMTF tracks, in same format as unpacked data
   produces<l1t::RegionalMuonCandBxCollection>("EMTF");  // EMTF tracks output to uGMT
+  produces<l1t::EMTFShowerCollection>();
```
Might be nice to also give this a label.

dildick replied:
Perhaps ```produces<l1t::ShowerCollection>("EMTF");``` is more suitable

dinyar commented on DataFormats/L1TMuon/interface/EMTFShower.h:
```diff
@@ -0,0 +1,32 @@
+
+namespace l1t {
+
+  class EMTFShower {
```
Thinking about this some more, I'm actually not sure it's a good idea to have xMTFShowers. In the end this all goes to the uGT and ideally we don't want to treat showers from various areas of the detector differently. Might be better to model this data format after RegionalMuonCand?

dildick replied:
Right, ultimately, it all comes together into the uGT. So perhaps like you suggest something more general. I can change it to ```l1t::Shower``` or ```l1t::RegionalShower``` (since it's per sector). Not sure if we would want to call it ```l1t::RegionalMuonShower``` since we're trying to capture hadronic showers and similar things, but that don't look at all like muons.

dildick commented:
Hi @rekovic @dinyar, do you have any comments/questions/suggestions so far?

dinyar commented:
Hi @dildick,

did you see my comments above?

Cheers,
Dinyar

dildick commented:
I'm afraid not...

dinyar commented:
@dildick  Ah, sorry hadn't pressed the submit button :)

jiafulow commented:
Hi Sven,

Thanks for preparing this. A few comments:

- I think it would be good to involve Andrew and Efe on this development.
- When you prepare the PR to official CMSSW, please split into separate PRs for CSC local trigger changes and EMTF changes.
- The EMTFTrack class is an unofficial class in the sense that it's not used by anyone downstream except for debugging purposes by EMTF people. I don't think there is a need for the EMTFShower class. What you would want is an input format for the uGMT. So, as Dinyar suggested, it's better to create a class based on RegionalMuonCand, and the particular class should belong to uGMT, not EMTF.
- The class "Shower" should have a more descriptive name like MuonShower (or HMT?), since you are putting it under the L1Trigger package. But I'm not sure why you would need that class? I thought the shower info is passed along by the LCT objects?
- In general, if a class or a function is simply a placeholder, please leave a comment stating that or put an FIXME tag.

Cheers,
Jia Fu

dildick commented:
@eyigitba @abrinke1 

Hi Jia Fu, thanks for chiming in. Yes, when the time comes, I'll split it up in CSC local trigger and EMTF. So it seems like you are also in favor of having an object that is more geared towards the uGMT than the EMTF. That is fine, and perhaps more appropriate for Run-4, where most likely we will have shower objects also in the barrel and the overlap regions. Are you suggesting something that derives from ```RegionalMuonCand```, e.g. ```RegionalMuonShower : public RegionalMuonCand```? So we do need an object because somehow we need to carry the shower information from the track-finders to the uGMT. LCTs do not make it beyond the track-finders. A dedicated object is one way to do this. I think in a L1 trigger meeting Vladimir suggested to define a new object.  Another way is to separate the bits over the 3 RegionalMuonCands per EMTF sector, but that is pretty messy  in the simulation - although this is what we are pursuing in the hardware.

jiafulow commented:
Hi Sven, Ok. No, I'm not suggesting `RegionalMuonShower : public RegionalMuonCand`, but I was just agreeing with Dinyar about "Might be better to model this data format after RegionalMuonCand?"

I agree that there needs to be a dedicated object from EMTF to uGMT, which I believe is why you created EMTFShower. But what about this guy? DataFormats/L1Trigger/interface/Shower.h 

Regarding your last point, I'd also prefer one object rather than 3, if it is just one piece of info, not 3 pieces, even though it gets chopped into 3 pieces in the DAQ format.

dildick commented:
Right, I did consider the fact that shower data from ETMF to uGMT might be different than data from uGMT to uGT. In the uGMT you probably want to do a few operations (e.g. global OR of showers in both endcaps) so that in the end you are only transmitting 1 or 2 bits from the uGMT to the uGT.  E.g. they could mean "there is at least one shower", "there is at least one shower in X stations", "there are 2 small showers",... That is why I wanted the distinction between EMTFShower and l1::Shower. Maybe I should call them `l1t::RegionalMuonShower` and ```l1t::MuonShower``` and put them both in DataFormats/L1Ttrigger

Track-finders produce
`l1t::RegionalMuonShower("BMTF")` Phase-2 ??
`l1t::RegionalMuonShower("OMTF")` Phase-2 ??   
`l1t::RegionalMuonShower("EMTF")` Run-3: 2 bits per BX per sector

uGMT combines them into 
`l1t::MuonShower` ~2 bits per BX

uGT receives `l1t::MuonShower`

jiafulow commented:
Ok, I see. Sorry I just realized you explained about "l1t::Shower" in the original post, but I missed that. A common object for the track finders sounds reasonable to me.

dildick commented:
@dinyar @rekovic @jiafulow I'll try to incorporate your comments into a new draft for the data formats. If I find time, I'll try to work out the logic that we need in the EMTF processors.

abrinke1 commented:
Hi @dildick ,

Sorry to jump in on this late.  I generally agree with the comments from @dinyar and @jiafulow .  A few quick thoughts:
* I think "l1t::MuonShower" might be more appropriate than "l1t::Shower", since other groups are working on displaced jet triggers in L1T, for example.
* If a data format is created to parallel "EMTFTrack", it should be created and implemented by the EMTF group, and should be output by the EMTF emulator.
* As stated by others, you need a new object to parallel the "RegionalMuonCand" and "MuonCand" classes, though it may be better for the uGMT and uGT folks to write these classes, or at least sign off on them.
* By way of reminder, the new shower bits appear inside a "CSCCorrelatedLCTDigi" object to be transmitted to EMTF?  And I guess we'll need some logic in the EMTF emulator to distinguish between:
  a) Regular LCTs with no shower.
  b) Regular LCTs which also have a shower.
  c) No regular LCT but with a shower.
I assume the new LCT format makes it straightforward to distinguish between "a", "b", and "c".

That's all for now.  I hope you have already presented this PR in the L1T Offline SW meeting, or will soon.

Best regards,
Andrew

dildick commented:
Hi @abrinke1, 

>> I think "l1t::MuonShower" might be more appropriate than "l1t::Shower", since other groups are working on displaced jet triggers in L1T, for example.

Done

>> If a data format is created to parallel "EMTFTrack", it should be created and implemented by the EMTF group, and should be output by the EMTF emulator.

I don't think we need this. It would be a bit overkill since the muon shower reconstruction is quite simple in the EMTF.

>> As stated by others, you need a new object to parallel the "RegionalMuonCand" and "MuonCand" classes, though it may be better for the uGMT and uGT folks to write these classes, or at least sign off on them.

Yep. Implemented a draft. To be presented to uGMT and uGT people

>> By way of reminder, the new shower bits appear inside a "CSCCorrelatedLCTDigi" object to be transmitted to EMTF? And I guess we'll need some logic in the EMTF emulator to distinguish between:
    a) Regular LCTs with no shower.
    b) Regular LCTs which also have a shower.
    c) No regular LCT but with a shower.
    I assume the new LCT format makes it straightforward to distinguish between "a", "b", and "c".

Case a) and b): in this case you can access the hmt member in the LCT. Case c)? For those I may need to implement a local shower object. It would be a weird case I guess, where there are at least 40 hits in a chamber, but collectively they do not fit a single LCT pattern...

abrinke1 commented:
Hi @dildick ,
Sounds good.  For case "(c)", i.e. shower object but no LCT, you or someone on your team could do the study: what percentage of signal showers (even "loose" showers) have 0 "good" LCTs transmitted from that CSC sector to EMTF?  If the answer is < 1%, the best solution is probably to simply make the presence of at least one "good" LCT an explicit requirement for the CSCs to transmit (or for EMTF to accept) a CSC shower object.  That way we avoid creating and keeping track of a new data format for some weird corner case, and it may also provide some protection against random CSC noise, e.g. noisy strips with no noise in wires, or visa versa.
